pyintegration/deephaven2/dtypes.py

```diff
@@ -99,31 +75,226 @@ def array_from(self, values: Iterable):
             DHError
         """
         try:
-            return jpy.array(self._j_name, values)
+            return jpy.array(self.j_name, seq)
         except Exception as e:
             raise DHError("failed to create a Java array.") from e
 
 
-bool_ = DType.bool_
-byte = DType.byte
-int8 = DType.int8
-short = DType.short
-int16 = DType.int16
-char = DType.char
-int_ = DType.int_
-int32 = DType.int32
-long = DType.long
-int64 = DType.int64
-float_ = DType.float_
-single = DType.single
-float32 = DType.float32
-double = DType.double
-float64 = DType.float64
-string = DType.string
-BigDecimal = DType.BigDecimal
-StringSet = DType.StringSet
-DateTime = DType.DateTime
-Period = DType.Period
+class IntegerDType(DType):
+    def __init__(self, qst_type, j_name, np_dtypes: Tuple = (), null_value=None):
+        super().__init__(qst_type=qst_type, j_name=j_name, is_primitive=True)
+        self.np_types = np_dtypes
+        self.null_value = null_value
+
+    def array_of(self, seq: Sequence):
+        """ Creates a Java array of the same integer type populated with values from a Python sequence.
+
+        Args:
+            seq: a Python sequence of compatible data type, can be a np array or Pandas series
+
+        Returns:
+            a Java array
+
+        Raises:
+            DHError
+        """
+        try:
+            if isinstance(seq, np.ndarray):
+                if seq.dtype in self.np_types:
+                    return super().array_of(seq)
+                elif np.issubdtype(seq.dtype, np.integer):
+                    return super().array_of(seq.astype(self.np_types[0]))
+                elif np.issubdtype(seq.dtype, np.floating):
+                    nan_arr = np.isnan(seq)
+                    arr = seq.astype(self.np_types[0])
+                    arr[nan_arr] = self.null_value
+                    return super().array_of(arr)
```

Comment:
1) Need to very carefully think through the casting consequences here.  Precision can be lost, as well as other nastiness.  What about `NULL_*` values?  Do we preserve them in the cast?  Ignore them?  Needs careful consideration.
2) If we are casting, we should document it, so that users know precision may be lost or other things.
3) Line 110 is mapping nan->null.  This should be documented.  We might also want a flag for how to handle this.  e.g. nan->null, exception, or whatever.
4) Should casting be controlled by an input argument?
5) We are only doing explicit casting for numpy types.  What happens for native Python types?

As you can tell, this section gives me the willies, so we should be very thoughtful in what is going on.

Reply:
latest thoughts:
1. always do unsafe casting but warn the users in the pydoc
2. for Integer types, it makes sense to always convert a NaN into the corresponding NULL value, document it
2. for floating types, by default no conversion is made for NaN, but array_of has an additional parameter to  request for NULL conversion

@chipkent what do you think?